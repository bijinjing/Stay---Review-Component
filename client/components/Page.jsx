import React from 'react';


class Page extends React.Component{
  constructor(props) {
    super(props)

    this.state={
      currentPage:1,
      pageHead:1
    }

  }


  render(){
    return(
      <form>
          {this.state.currentPage !== 1 && <input id="left" value="left" type="button"/>}
          <input value='1' type="button"/>
          {[1,2,3].includes(this.state.currentPage) && <input value={this.state.pageHead} type="button"/>}
          {[2,3,4].includes(this.state.currentPage) && <input value={this.state.pageHead+1} type="button"/>}
          <input value={this.state.pageHead+2} type="button"/>
          <input value={this.props.lastPage-2} type="button"/>
          <input value={this.props.lastPage-1} type="button"/>

          <input value={this.props.lastPage} type="button"/>
          {this.state.currentPage !== this.props.lastPage && <input id="right" value="right" type="button"/>}
      </form>
    )
  }
}

export default Page


