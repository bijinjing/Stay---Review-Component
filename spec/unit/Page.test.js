import React from 'react';
import Page from '../../client/components/Page.jsx';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import Enzyme from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import {configure} from 'enzyme';
import {shallow, mount, render} from 'enzyme';

configure({ adapter: new Adapter() });

describe('Page', ()=>{

  it('renders without crashing', () => {
    const props = {
      currentPage:3,
      lastPage:10,
    }

    const wrapper = shallow(<Page {...props}/>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('page inputs are renderred properly when current page is 1 ', () => {

    const props = {
      currentPage:1,
      lastPage:10,
    }

    const component = mount(<Page {...props}/>)
    expect(component.find('#left').length).toEqual(0);
    expect(component.find('#right').length).toEqual(1);
    expect(component.find('.pg').length).toEqual(4);

    component.unmount();

  })

  it('page inputs are renderred properly when current page is 2 ', () => {
    const props = {
      currentPage:2,
      lastPage:10,
    }

    const component = mount(<Page {...props}/>)
    expect(component.find('#left').length).toEqual(1);
    expect(component.find('#right').length).toEqual(1);
    expect(component.find('.pg').length).toEqual(4);

    component.unmount();

  })

  it('page inputs are renderred properly when current page is 3 ', () => {
    const props = {
      currentPage:3,
      lastPage:10,
    }

    const component = mount(<Page {...props}/>)
    expect(component.find('#left').length).toEqual(1);
    expect(component.find('#right').length).toEqual(1);
    expect(component.find('.pg').length).toEqual(5);

    component.unmount();

  })

  it('page inputs are renderred properly when current page is 4 ', () => {
    const props = {
      currentPage:4,
      lastPage:10,
    }

    const component = mount(<Page {...props}/>)
    expect(component.find('#left').length).toEqual(1);
    expect(component.find('#right').length).toEqual(1);
    expect(component.find('.pg').length).toEqual(6);

    component.unmount();

  })

  it('page inputs are renderred properly when current page is in the middle ', () => {
    const props = {
      currentPage:5,
      lastPage:10,
    }

    const component = mount(<Page {...props}/>)
    expect(component.find('#left').length).toEqual(1);
    expect(component.find('#right').length).toEqual(1);
    expect(component.find('.pg').length).toEqual(5);

    component.unmount();

  })

  it('page inputs are renderred properly when current page is last page ', () => {
    const props = {
      currentPage:10,
      lastPage:10,
    }

    const component = mount(<Page {...props}/>)
    expect(component.find('#left').length).toEqual(1);
    expect(component.find('#right').length).toEqual(0);
    expect(component.find('.pg').length).toEqual(4);

    component.unmount();

  })

  it('page inputs are renderred properly when current page is second last page ', () => {
    const props = {
      currentPage:9,
      lastPage:10,
    }

    const component = mount(<Page {...props}/>)
    expect(component.find('#left').length).toEqual(1);
    expect(component.find('#right').length).toEqual(1);
    expect(component.find('.pg').length).toEqual(4);

    component.unmount();

  })

  it('page inputs are renderred properly when current page is third last page ', () => {
    const props = {
      currentPage:8,
      lastPage:10,
    }

    const component = mount(<Page {...props}/>)
    expect(component.find('#left').length).toEqual(1);
    expect(component.find('#right').length).toEqual(1);
    expect(component.find('.pg').length).toEqual(5);

    component.unmount();

  })

  it('page inputs are renderred properly when current page is forth last page ', () => {
    const props = {
      currentPage:7,
      lastPage:10,
    }

    const component = mount(<Page {...props}/>)
    expect(component.find('#left').length).toEqual(1);
    expect(component.find('#right').length).toEqual(1);
    expect(component.find('.pg').length).toEqual(6);

    component.unmount();

  })

  it('page inputs are renderred properly when pages are less than 5 ', () => {
    const props = {
      currentPage:2,
      lastPage:4,
    }

    const component = mount(<Page {...props}/>)
    expect(component.find('#left').length).toEqual(1);
    expect(component.find('#right').length).toEqual(1);
    expect(component.find('.pg').length).toEqual(4);

    component.unmount();

  })

  it('changeCurrentPage and renderCurrentPage are invoked when right button is clicked', () => {
    let current = true;
    let render = true
    const props = {
      currentPage:3,
      lastPage:10,
      changeCurrentPage:() => {current = false},
      renderCurrentPage:() => {render = false}
    }

    const component = mount(<Page {...props}/>)
    component
      .find('#right')
      .simulate('click')

    expect(current).toEqual(false);
    expect(render).toEqual(false);

    component.unmount();

  })

  it('changeCurrentPage and renderCurrentPage are invoked when left button is clicked', () => {
    let current = true;
    let render = true
    const props = {
      currentPage:3,
      lastPage:10,
      changeCurrentPage:() => {current = false},
      renderCurrentPage:() => {render = false}
    }

    const component = mount(<Page {...props}/>)
    component
      .find('#left')
      .simulate('click')

    expect(current).toEqual(false);
    expect(render).toEqual(false);

    component.unmount();
  })

  it('changeCurrentPage and renderCurrentPage are invoked when page number button is clicked', () => {
    let current = true;
    let render = true
    const props = {
      currentPage:3,
      lastPage:10,
      changeCurrentPage:() => {current = false},
      renderCurrentPage:() => {render = false}
    }

    const component = mount(<Page {...props}/>)
    component
      .find('.pg')
      .first()
      .simulate('click')

    expect(current).toEqual(false);
    expect(render).toEqual(false);

    component.unmount();
  })
})