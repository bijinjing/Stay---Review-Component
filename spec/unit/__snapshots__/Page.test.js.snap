// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page renders without crashing 1`] = `
<form>
  <input
    id="left"
    onClick={[Function]}
    type="button"
    value="left"
  />
  <input
    className="pg"
    onClick={[Function]}
    type="button"
    value={1}
  />
  <input
    className="pg"
    onClick={[Function]}
    type="button"
    value={2}
  />
  <input
    className="pg"
    onClick={[Function]}
    type="button"
    value={3}
  />
  <input
    className="pg"
    onClick={[Function]}
    type="button"
    value={4}
  />
  <a>
    ...
  </a>
  <input
    className="pg"
    onClick={[Function]}
    type="button"
    value={10}
  />
  <input
    id="right"
    onClick={[Function]}
    type="button"
    value="right"
  />
</form>
`;
